x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/JobModelSnapshotUpgrader.java

```diff
@@ -0,0 +1,308 @@
+            // time field
+            fieldIndexes.put(job.getDataDescription().getTimeField(), 0);
+            int index = 1;
+            for (String field : job.getAnalysisConfig().analysisFields()) {
+                if (AnalysisConfig.ML_CATEGORY_FIELD.equals(field) == false) {
+                    fieldIndexes.put(field, index++);
+                }
+            }
+            fieldIndexes.put(LengthEncodedWriter.CONTROL_FIELD_NAME, index);
+            return fieldIndexes;
+        }
+
+        void writeHeader() throws IOException {
+            Map<String, Integer> outFieldIndexes = outputFieldIndexes();
+            // header is all the analysis input fields + the time field + control field
+            int numFields = outFieldIndexes.size();
+            String[] record = new String[numFields];
+            for (Map.Entry<String, Integer> entry : outFieldIndexes.entrySet()) {
+                record[entry.getValue()] = entry.getKey();
+            }
+            // Write the header
+            process.writeRecord(record);
+        }
+
+        void restoreState() {
+            try {
+                process.restoreState(stateStreamer, params.modelSnapshot());
+            } catch (Exception e) {
+                logger.error(() -> new ParameterizedMessage("[{}] [{}] failed to write old state", jobId, snapshotId), e);
+                setTaskToFailed("Failed to write old state due to: " + e.getMessage(),
+                    ActionListener.wrap(t -> shutdown(e), f -> shutdown(e)));
+                return;
+            }
+            task.updatePersistentTaskState(
+                new SnapshotUpgradeTaskState(SnapshotUpgradeState.READING_NEW_STATE, task.getAllocationId(), ""),
+                ActionListener.wrap(
+                    readingNewState -> {
+                        if (continueRunning.get() == false) {
+                            shutdown(null);
+                            return;
+                        }
+                        submitOperation(() -> {
+                            writeHeader();
+                            process.persistState(
+                                // C++ is expecting the timestamp to be in seconds, not Milliseconds
+                                params.modelSnapshot().getTimestamp().getTime()/1000,
+                                params.modelSnapshot().getSnapshotId(),
+                                params.modelSnapshot().getDescription());
+                            return null;
+                            // Execute callback in the UTILITY thread pool, as the current thread in the callback will be one in the
+                            // autodetectWorkerExecutor. Trying to run the callback in that executor will cause a dead lock as that
+                            // executor has a single processing queue.
+                        }, (aVoid, e) -> threadPool.executor(UTILITY_THREAD_POOL_NAME).execute(() -> shutdown(e)));
```

Comment:
The regular autodetect communicator doesn't have this issue as the callbacks don't ever feed back into the same executor pool. But here, it does, and to handle it, I am pushing off to the utility thread to execute the shutdown (which needs to use the autodetect executor).